tweaks to mailer

andrewplummer opened the pull request:
A few tweaks to the mailer util:

- Allow a [front matter](https://www.npmjs.com/package/front-matter) at the beginning of HTML templates to allow them to define their own subject lines (and potentially more metadata)
- Allow any environment key in email templates and normalized env naming.
- Fallback for text based emails when they don't exist. Text emails will strip out html tags and additionally add links in the format `<a href="href">text</a>` -> `text (href)`.
- Allow `sendTemplatedEmail` to simply accept a user object and correctly construct the target user's email. This can still be overridden using `to`.
- Env config to allow postmark sending in development for debugging.
- Fixed various escaping issues.

kaareal commented on services/api/README.md:
```diff
@@ -65,15 +65,15 @@ All configuration is done using environment variables. The default values in `.e
 - `APP_URL` - URL for app defaults to `http://localhost:2200`
 - `POSTMARK_FROM` - Reply email address `no-reply@bedrock.foundation`
-- `POSTMARK_APIKEY` - APIKey for Postmark `[change me]`
+- `POSTMARK_API_KEY` - APIKey for Postmark `[change me]`
```
can we lose the [change me] seems stupid now

andrewplummer replied:
ah yeah, totally :)

kaareal commented on services/api/emails/layout.html:
```diff
@@ -314,14 +313,13 @@
                   <tr>
                     <td class="content-cell" align="center">
-                      <p class="f-fallback sub align-center">&copy; 2021 {{appName}}. All rights reserved.</p>
+                      <p class="f-fallback sub align-center">&copy; 2021 {{APP_NAME}}. All rights reserved.</p>
```
lets generate the year (2021) so we don't have to update this every year

andrewplummer replied:
ah yeah good idea... was thinking about that too

kaareal commented on services/api/src/utils/mailer.js:
```diff
@@ -1,15 +1,107 @@
 
-const POSTMARK_FROM = config.get('POSTMARK_FROM');
-const ENV_NAME = config.get('ENV_NAME');
+const ENV = config.getAll();
```
better to ask for them individually ... i am wondering if you trying get around something here?

andrewplummer replied:
why is that? I want all configs to be available to the mailers by their exact name in .env... not doing this is unexpected and increases friction IMO

kaareal replied:
The point bedrock/config was to avoid the case where you running in production with env vars that are missing, but only finding out about very late ... as nothing warns you unless you are trying to access the env.

It's not a lot of fiction though? 
We are looking at list of 2-3 env vars no?

Also feel uncomfortable with having all the "secrets" exposed with an bad email.

andrewplummer replied:
I don't consider secrets to be "exposed" any more than they are to the rest of the code base which also has access.

> running in production with env vars that are missing

I don't think this is true right? As environment variables are inherited they would have to somehow be manually removed, I don't think that's possible.

As a use case we added a "HELP_DESK_URL" to the env. There are untold numbers of variables that might be added so keeping that in sync with the mailer is a lot of friction, yes, especially as the names were different for some vars.

kaareal replied:
`I don't consider secrets to be "exposed" any more than they are to the rest of the code base which also has access.`
My thinking was that you might have dynamic templates in the near future.

```I don't think this is true right? As environment variables are inherited they would have to somehow be manually removed, I don't think that's possible.
```
Thats assuming there is a default, that is not the case always. Specially as we started to move more things into secrets. I am start to think bedrock/config is not solving that problem ... specially how it's being used right now. But let's ignore that, i think i need to rethink bedrock/config. 

```
As a use case we added a "HELP_DESK_URL" to the env. There are untold numbers of variables that might be added so keeping that in sync with the mailer is a lot of friction, yes, especially as the names were different for some vars.
```
Fair enough, i wish we could do like for the web ... we we have a prefix. 

But fine doesn't seems like there is any easy solution. And we are not doing dynamic templates yet so not a worry.

andrewplummer replied:
> My thinking was that you might have dynamic templates in the near future.

Why would that be an issue?

> Fair enough, i wish we could do like for the web ... we we have a prefix.

We do? I don't see one in master.

Actually I like your first idea better... env configs are ALWAYS public and if they are not then we should put them in secrets instead.

kaareal replied:
thats interesting but it all gets merged... but perhaps there is a way to split them apart

andrewplummer replied:
ie. anything private would be marked `PRIVATE_` (or something)... that would serve as a signal that it should be inside a k8s secret config (easy to see when in the yaml files) ... however it could also easily allow `config.getPublic()` which could simply be `getAll` but excluding anything marked as private

kaareal replied:
i think there is something there ... but yeah that shouldnt hold this back

kaareal replied:
Just added @bedrockio/config (2.2.1) which now exposes getAllSafe, which returns all keys that has the prefix `SAFE_`

andrewplummer replied:
ok I don't think this should hold up this PR but fwiw I really think we should flip that. All .env configs should be assumed to be safe by default as we should never be checking in private configs.

Using the `PUBLIC_` scheme will mean we will have to add that prefix everywhere and adds a lot of friction.

kaareal replied:
how much fiction is there really in writing a prefix

andrewplummer replied:
A lot considering most everything should be exposed by default

kaareal replied:
do you have examples as this does not compute, by far in my experience env vars is for apis that should not be exposed

andrewplummer replied:
Actually you could be right...hard to know until we actually use it... let's go with this for now 👍

kaareal commented on services/api/src/utils/mailer.js:
```diff
@@ -1,15 +1,107 @@
+    return;
+  }
+
+  options = {
```
could we call this something else than `options`, redeclaring vars makes me confused

andrewplummer replied:
Maybe call it `vars` instead to drive home that the variables are interpolated?

kaareal commented on services/api/src/utils/mailer.js:
```diff
@@ -30,90 +122,70 @@ marked.use({
+// which causes issues so avoid this.
+const escapeHtml = Mustache.escape;
+Mustache.escape = (text) => {
+  if (!text.match(/https?:\/\//)) {
```
love it

andrewplummer replied:
hah. I thought you'd hate it... only after I did this I realized that `{{&something}}` and `{{{something not escaped}}}` were a thing and I forgot about them... that's the "mustache" way to do it, however the fact that I forgot makes me want to keep it... I don't think I trust even myself to always remember this and then on top of that I can easily see non-technical people to be writing email templates (we tend to go nuts with this) and having to teach them that seems silly... feels hacky but I'd like to keep it

kaareal commented on services/api/src/utils/mailer.js:
```diff
@@ -30,90 +122,70 @@ marked.use({
+  if (body) {
+    return Mustache.render(body, options);
+  } else {
+    return '';
```
what the case for not having a body?

andrewplummer replied:
Mostly because I was unsure of the implication of sending `null` or `undefined` to postmark as part of the html or text (might get stringified, you'd get a full html email but the text version would just say "undefined" or something) so I thought I'd normalize to an empty string here as a precaution.

kaareal replied:
make sense

kaareal commented on services/api/src/utils/mailer.js:
```diff
@@ -1,15 +1,107 @@
+}
+
+async function sendMail({ to, subject, html, text, body, options }) {
+  if (canSendMail()) {
```
don't love this construct, wondering if that could be written simpler where you return early.
i love return early, think thats generally the correct pattern.

andrewplummer replied:
100% disagree :) ... there are some cases where returning up front is more readable but if/else exists for a reason I much prefer it in common cases.

andrewplummer replied:
besides in this case what would it be like:

```
if (canSendMail()) {
  // do the main thing
}
// do the non-main thing as a fallback? that's almost reversed from what you'd expect
```

readability is subjective but it's also situational

kaareal replied:
lets leave that behind ... as you said `readability is subjective`

andrewplummer replied:
kk how about this then... inlined everything and instead made `sendEmail` actually do what it claims it does

kaareal commented on services/api/src/utils/mailer.js:
```diff
@@ -34,8 +34,12 @@ async function sendTemplatedMail({ template, layout = 'layout.html', to, ...opti
+  // to also contain variables so that dynamic contenet can work.
+  let body = templateBody;
+  body = interpolate(body, vars);
+  body = interpolate(body, vars);
```
mmm that smells like something isnt right the ... cant that lead to problems ? 
Do you have an example ? of what you try to solve here

andrewplummer replied:
this is a first step to allow dynamic emails that can be edited by admins. Having a 2 pass system allows both the email templates as well as the injected comments to reference variables.

kaareal replied:
https://github.com/bedrockio/bedrock-core/pull/169#discussion_r764920727

andrewplummer replied:
it's not really an issue yet as our use case isn't fleshed out... dynamic doesn't mean the source isn't trusted. we can cross that bridge when we come to it

andrewplummer commented:
@kaareal ok did some more refactoring here. Removed the logic for double-pass interpolation. Instead the files are loaded (and cached) and interpolation happens on the template body as it is passed in. This means that there are two signatures for `sendTemplatedMail`:

```js
sendTemplatedMail({ file: 'welcome.md' });
```

```js
sendTemplatedMail({ template: 'hello from a dynamic template body' });
```

Additionally both file based and dynamic templates can describe an optional "front matter" describing metadata including their subject:

```js
sendTemplatedMail({
  template: `
---
subject: 'I am the subject and can have {{dynamic}} variables as well'
---

I'm the mail body.
```

Note that in the project I'm using this in I've added an "EmailTemplateField" input that handles parsing of dynamic templates and separates into subject/body fields and works transparently as any other form field, however we're not actually using dynamic templates in bedrock yet so I've left that out for now.

kaareal commented on services/api/src/emails/layout.html:
```diff
@@ -281,8 +286,8 @@
-                <a href="{{&appUrl}}" class="f-fallback email-masthead_name">
-                  <img src="{{&appLogoUrl}}" alt="{{appName}}" />
+                <a href="{{&APP_URL}}" class="f-fallback email-masthead_name">
+                  <img src="{{&APP_LOGO_URL}}" alt="{{APP_NAME}}" class="email-masthead_logo" />
```
is the `&` still needed ?

andrewplummer replied:
perhaps not you can do either way... I'm fine to remove if you prefer

andrewplummer commented on services/api/.env:
```diff
@@ -17,7 +17,8 @@ JWT_SECRET=m6bv7!m|E:7:*X3Mx7I)2+Y>B70+&K
 POSTMARK_FROM=no-reply@bedrock.foundation
-POSTMARK_APIKEY=
+POSTMARK_API_KEY=
+POSTMARK_DEV_TO=
```
Just a shortcut so that you don't have to do that. It's quite common to debug emails. This does apply to postmark too because it only applies to `Postmark.ServerClient` ... we don't really have any other mailer

kaareal replied:
I would say this is business logic of mailer and not postmark ... we calling the wrapper "mailer" because we are not bound to postmark ... can we easy be swapped ... recall AD with blaze or what ever it was called.

andrewplummer replied:
Then why is it `POSTMARK_FROM` and not `MAILER_FROM`? I'm fine with being accurate but you have to weight it against path of least surprise where the question will be "so what's the difference between "postmark" and "mailer"?" when in fact there is none

andrewplummer commented on services/api/jest.config.js:
```diff
@@ -6,4 +6,7 @@ module.exports = {
   // https://github.com/shelfio/jest-mongodb#6-jest-watch-mode-gotcha
   watchPathIgnorePatterns: ['globalConfig'],
+  moduleNameMapper: {
+    postmark: '<rootDir>/src/utils/testing/mocks/postmark',
```
well this is mocking the actual postmark npm package itself for testing

kaareal replied:
then why are you mocking the mailer's sendTemplatedEmail?

andrewplummer replied:
ah I'm actually not anymore...sorry that was part of the refactor, should have mentioned
